### PR TITLE
Merge New Features into Development

### PR DESCRIPTION
**Pull request by EngineerSuite2**

### CODE REVIEW

File: FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        public PreviewNPCViewModel(NPCModel nPCModel)
+        {
+            NPCModel = nPCModel;
+            SpeedDescription = UpdateSpeedDescription();
+            SkillsDescription = _updateSkillsDescription();
+        }
+
+        public string UpdateSpeedDescription()
+        {
+            StringBuilder sl = new StringBuilder();
```
Comment (lorddusk) [inline FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs:37]: sl > stringBuilder

File: FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        private string _appendSpeedAttribute(string name, int value, bool hover)
+        {
+            if (value != 0)
+                return name + " " + value + " ft." + (hover ? " (hover), ": ", ");
+            return "";
+        }
+
+        private string _updateSkillsDescription()
+        {
+            StringBuilder sb = new StringBuilder();
```
Comment (lorddusk) [inline FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs:61]: sb > stringBuilder

File: FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        private void ESExit_Click(object sender, RoutedEventArgs e)
+        {
+            Close();
+        }
+
+        private void UpdateReaction_Click(object sender, RoutedEventArgs e)
+        {
+            ActionModelBase lam = new ActionModelBase();
```
Comment (lorddusk) [inline FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs:48]: Use proper variable names, I have no idea what lam is supposed to mean?

File: FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        private void OverviewControl_RemoveAction(object sender, EventArgs e)
+        {
+            actionController.RemoveReaction((sender as OverviewControl).DataContext as ActionModelBase);
+        }
+
+        private void OverviewControl_EditAction(object sender, EventArgs e)
+        {
+            ActionModelBase temp = (sender as OverviewControl).DataContext as ActionModelBase;
```
Comment (lorddusk) [inline FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs:61]: see comment above.

File: FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                    return (action as LegendaryActionModel);
+                }
+            }
+            return null;
+        }
+
+        private void LegendaryActionOverviewControl_RaiseActionInList(object sender, EventArgs e)
+        {
+            var model = getModelFromDataContext(sender);
+            if(model != null)
```
Comment (lorddusk) [inline FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs:76]:
```suggestion
            if (model != null)
```

File: FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        private void LegendaryActionOverviewControl_EditAction(object sender, EventArgs e)
+        {
+            var model = getModelFromDataContext(sender);
+            if (model.ActionName.Equals(LegendaryActionOptionControl.ActionName))
+            {
+                OptionsRadioButton.IsChecked = true;
+                stackOptions.Visibility = Visibility.Visible;
+                stackActions.Visibility = Visibility.Hidden;
+                legendaryOptionsControl.DataContext = model;
+            } else
```
Comment (lorddusk) [inline FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs:104]:
```suggestion
            } 
            else
```

File: FMPTests/NPC/ViewModels/PreviewNPCViewModelTests.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        [DataRow(30, 30, 30, 0, false, 0, "30 ft., climb 30 ft., burrow 30 ft.")]
+        [DataRow(30, 30, 0, 30, false, 0, "30 ft., fly 30 ft., burrow 30 ft.")]
+        [DataRow(30, 30, 0, 30, true, 0, "30 ft., fly 30 ft. (hover), burrow 30 ft.")]
+        [DataRow(30, 30, 0, 0, false, 30, "30 ft., burrow 30 ft., swim 30 ft.")]
+        [DataRow(30, 30, 30, 30, false, 0, "30 ft., climb 30 ft., fly 30 ft., burrow 30 ft.")]
+        [DataRow(30, 30, 30, 30, true, 0, "30 ft., climb 30 ft., fly 30 ft. (hover), burrow 30 ft.")]
+        [DataRow(30, 30, 30, 0, false, 30, "30 ft., climb 30 ft., burrow 30 ft., swim 30 ft.")]
+        [DataRow(30, 30, 30, 30, false, 30, "30 ft., climb 30 ft., fly 30 ft., burrow 30 ft., swim 30 ft.")]
+        [DataRow(30, 30, 30, 30, true, 30,  "30 ft., climb 30 ft., fly 30 ft. (hover), burrow 30 ft., swim 30 ft.")]
+        public void UpdateSpeedDescriptionTest(int speed, int burrow, int climb, int fly, bool hover, int swim, string expected)
```
[the remark this answers is not in the document]
Comment (EngineerSuite2) [reply FMPTests/NPC/ViewModels/PreviewNPCViewModelTests.cs:40]: That was in Testing. It isn't truly implemented anywhere. I could honestly delete the whole section.

[the remark this answers is not in the document]
Comment (fredska) [reply FMPTests/NPC/ViewModels/PreviewNPCViewModelTests.cs:40]: I'd rather not get rid of an applicable unit test.  I'm up for suggestions of differently formatted parameterized tests, if there are any.

File: FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs (hunk 6 above)
Comment (fredska) [reply FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs:104]: Java vs .NET habits it seems :)

File: FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs (hunk 5 above)
Comment (EngineerSuite2) [reply FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs:76]: I think this is a file that doesn't exist anymore.

File: FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs (hunk 6 above)
Comment (EngineerSuite2) [reply FantasyModuleParser/NPC/Views/ViewLegendaryAction.xaml.cs:104]: i think this is a file that doesn't exist anymore.

File: FMPTests/NPC/ViewModels/PreviewNPCViewModelTests.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (EngineerSuite2) [reply FMPTests/NPC/ViewModels/PreviewNPCViewModelTests.cs:40]: We shall leave as-is at this point.

File: FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs (hunk 1 above)
Comment (EngineerSuite2) [reply FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs:37]: I don't understand the comment. Are you asking for a change?

File: FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs (hunk 2 above)
Comment (EngineerSuite2) [reply FantasyModuleParser/NPC/ViewModels/PreviewNPCViewModel.cs:61]: I don't understand the comment. Are you asking for a change?

File: FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs (hunk 3 above)
Comment (EngineerSuite2) [reply FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs]: Updated variables

File: FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs (hunk 4 above)
Comment (EngineerSuite2) [reply FantasyModuleParser/NPC/Views/Action/Reaction/Reactions.xaml.cs]: updated variables

**Comment by EngineerSuite2** [comment]
All requested changes have been performed

**Pull request merged**